### PR TITLE
Vuhi code review 2

### PR DESCRIPTION
**Pull request by vuhi**
- Add example of unit tests for Controller + DTO
- Add constants in Controller for consistent unit test
- Add additional Exception throw for Service layer

### **Analysis**
**Analysis of the program**
-There only a few minor changes since my last pull request. I expected this project should have gone further than the current state. The three major functionalities were not implemented (Random, Top3, Filter Search). Overall, the app worked (complied & serve).
-Was the program available in UC Github on time? YES
-Is the program documented/commented well enough for you to understand? YES
-Does the program compile? YES

**Learned Technical Concepts**
- Set Up + Unit test Normal Controller ( Return templates + view )
```
@RunWith(SpringRunner.class)
@SpringBootTest()
@AutoConfigureMockMvc
public class WhereToEatControllerTests {
    @Autowired
    private WhereToEatController controller;
...
```
- Set Up  + Unit test Normal Class ( DTO )
```
@RunWith(SpringRunner.class)
@SpringBootTest()
public class RestaurantDTOTests {
    private String expectName = "MOCK NAME";
    public RestaurantDTO initRestaurantDTO() {
        RestaurantDTO restaurantDTO = new RestaurantDTO();
        restaurantDTO.setName(expectName);
        restaurantDTO.setDescription("MOCK NAME");
        restaurantDTO.setAddress("MOCK NAME");
        restaurantDTO.setHours("MOCK NAME");
        restaurantDTO.setOpen(true);
        restaurantDTO.setWebsite("MOCK NAME");
        restaurantDTO.setRating(2.5);
        restaurantDTO.setPriceScale("MOCK NAME");
        restaurantDTO.setBysyness("MOCK NAME");
        restaurantDTO.setLatitude(100);
        restaurantDTO.setLongitude(200);
        restaurantDTO.setFoodType(FoodType.ASIAN);
        return restaurantDTO;
    }
...
```
- Mock request with `MockMvc`
```
 this.mockMvc.perform(get("/start"))
                .andExpect(status().isOk())
                .andExpect(view().name(WhereToEatController.START_TEMPLATE));
```

### CODE REVIEW

File: src/main/java/com/wheretoeat/WhereToEatController.java [hunk 1]
```diff
@@ -17,9 +17,14 @@
 import org.springframework.boot.SpringApplication;
 
 @Controller
-@SpringBootApplication
 public class WhereToEatController {
 
+	public static final String FOOD_ATTRIBUTE = "foodTypeOptions";
```
Comment (vuhi) [inline src/main/java/com/wheretoeat/WhereToEatController.java:22]: Declare a constant will help reading code easy + unit test

File: src/main/java/com/wheretoeat/WhereToEatController.java [hunk 2]
```diff
@@ -17,9 +17,14 @@
 import org.springframework.boot.SpringApplication;
 
 @Controller
-@SpringBootApplication
```
Comment (vuhi) [inline src/main/java/com/wheretoeat/WhereToEatController.java:20]: Unit test throws an error if there are more than `@SpringBootApplication` annotation

File: src/main/java/com/wheretoeat/service/IRestaurantService.java [hunk 3]
```diff
@@ -11,11 +11,11 @@
 	 * @param type - Takes a food type (category) enum 
 	 * @return will return a restaurant object based on the food type selected.
 	 */
-	RestaurantDTO fetchByFoodType(FoodType type);
+	RestaurantDTO fetchByFoodType(FoodType type) throws Exception;
```
Comment (vuhi) [inline src/main/java/com/wheretoeat/service/IRestaurantService.java:14]: Should handle Exception

File: src/test/java/com/wheretoeat/wheretoeat/RestaurantDTOTests.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package com.wheretoeat.wheretoeat;
```
Comment (vuhi) [inline src/test/java/com/wheretoeat/wheretoeat/RestaurantDTOTests.java:1]: This is just a unit test example, you can continue this format with other DTO. Or you could use that `@Before` annotation.

File: src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        try {
+            assertThat(controller).isNotNull();
+        } catch (Exception e) {
+            fail();
+        }
+    }
+
+    @Test
+    public void shouldHasView_start() throws Exception {
+        this.mockMvc.perform(get("/start"))
```
Comment (vuhi) [inline src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java:40]: mock `GET` request to `/start`

File: src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                .andExpect(status().isOk())
```
Comment (vuhi) [inline src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java:41]: Assert view, model, status ... etc

File: src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                .andExpect(view().name(WhereToEatController.START_TEMPLATE));
+    }
+
+    @Test
+    public void viewShouldHasFollowingAttributes() throws Exception {
+        this.mockMvc.perform(get("/start"))
+            .andExpect(model().attributeExists(
+                WhereToEatController.FOOD_ATTRIBUTE,
```
Comment (vuhi) [inline src/test/java/com/wheretoeat/wheretoeat/WhereToEatControllerTests.java:49]: The benefit of using variable is that you don't have to rewrite your test when you change the value in the controller. It is also very easy to read your code as well. This is my personal opinion. Feel free to twist it the way you like.

File: src/main/java/com/wheretoeat/service/IRestaurantService.java (hunk 3 above)
Comment (discospiff) [reply src/main/java/com/wheretoeat/service/IRestaurantService.java:14]: Does it actually throw an exception?  If not, leave this off.

**Comment by discospiff** [comment]
Several good changes here... and the discussion is quite good as well!
The group should consider merging these changes, or using this branch as a guide and implement them independently.

File: src/main/java/com/wheretoeat/service/IRestaurantService.java (hunk 3 above)
Comment (vuhi) [reply src/main/java/com/wheretoeat/service/IRestaurantService.java:14]: It should. In my experience, any services that handle querying the database always have the potential to throw DB Exception (Ex: Connection Exception).

**Pull request merged**